Improved the jest reporter with snapshot info per test.

excitement-engineer opened the pull request:
Improved the jest terminal reporter to show snapshot info per test. The snapshot summary shown at the end of the test run shows some generic info like *"2 snapshots written in 1 test suite"*, *"1 obsolete snapshot file found"*. However, it wasn't always immediately clear in what test suite these snapshot changes had occurred. This PR attempts to resolve this by showing all the snapshot changes per test file so that it is immediately clear what happened after a test run. I am curious to hear what you think!

This addresses issue #3581.

I was also thinking that it would be nice if the name of the snapshot which has been updated is shown in the terminal rather than simply stating *"2 snapshots written"* or *"2 obsolete snapshots found*". Adding the name to the output does not seem very trivial; the type `TestResult` used in the  jest-cli reporter doesn't export the names of the snapshots, it only exports the properties shown below. Therefore, adding the updated snapshot names would entail changing the structure in which the reporter gets its data. What do you guys think, is it worth adding the snapshot names to the output?

```
// the snapshot structure for type TestResult
snapshot: {|
    added: number,
    fileDeleted: boolean,
    matched: number,
    unchecked: number,
    unmatched: number,
    updated: number,
  |},
```

![jest-snapshot-demo 1](https://cloud.githubusercontent.com/assets/19558321/26486118/1754c932-41fa-11e7-872b-f279b540ce22.gif)

Let me know what you think!

orta commented:
The update to the TestResult is nice, then editors can run updates for specific tests too 👍

excitement-engineer commented:
Any updates on the status of this PR?

cpojer commented:
I'd like to merge this for 21, sorry for not getting to it earlier.

@excitement-engineer do you mind rebasing this change? Also, can we move it below console logs and failures, to the bottom of the result per test? :)

excitement-engineer commented:
Hey @cpojer, no problem! I refactored the code a bit to make it cleaner and I wrote some tests for the functionality I added. I also moved it below the console logs and failures:) Let me know what you think!

excitement-engineer commented:
I rebased the changes!

cpojer commented:
Lovely. Thanks for the PR! :)
